cocotb/binary.py

```diff
@@ -32,27 +32,59 @@
 
 import os
 import random
+import re
 import warnings
 
+
+RESOLVE_TO_0     = "-lL"
+RESOLVE_TO_1     = "hH"
+RESOLVE_TO_ERROR = "xXzZuUwW"  # Resolve to a ValueError() since these usually mean something is wrong
 resolve_x_to = os.getenv('COCOTB_RESOLVE_X', "VALUE_ERROR")
 
-def resolve(string):
-    for char in BinaryValue._resolve_to_0:
-        string = string.replace(char, "0")
-    for char in BinaryValue._resolve_to_1:
-        string = string.replace(char, "1")
-    for char in BinaryValue._resolve_to_error:
-        if resolve_x_to == "VALUE_ERROR" and char in string:
-            raise ValueError("Unable to resolve to binary >%s<" % string)
+class Resolver(dict):
+    """One-pass resolver for binary strings.
+
+    Sub-class :class:`dict` and use the key-value mapping for bit resolution.
+
+    Utilizes :meth:`re.sub()` to resolve all characters in one pass.
```

Comment:
You might want to look at using `string.maketrans` which will likely be even faster

Reply:
After investigating using `str.translate()`, it's much faster but it uses the translation value for all matching characters, so all `x` or `Z` become the same number.
The effect is the current behavior, where `xxxx` becomes `1111` or `0000`.

I can cleanup the `re.sub()` implementation if we really want per-bit random values.

Reply:
Since this patch is about performance, we should err on the side of keeping the current behavior.

In a future pass we could do a second pass to randomize X's.